src/main/java/com/minecolonies/coremod/client/render/worldevent/ColonyBlueprintRenderer.java

```diff
@@ -0,0 +1,381 @@
+import com.minecolonies.api.colony.buildings.ModBuildings;
+import com.minecolonies.api.colony.buildings.views.IBuildingView;
+import com.minecolonies.api.colony.workorders.IWorkOrderView;
+import com.minecolonies.api.colony.workorders.WorkOrderType;
+import com.minecolonies.api.items.ModItems;
+import com.minecolonies.api.tileentities.TileEntityColonyBuilding;
+import com.minecolonies.api.util.BlockPosUtil;
+import com.minecolonies.api.util.MathUtils;
+import com.minecolonies.coremod.colony.workorders.view.WorkOrderBuildingView;
+import net.minecraft.client.Minecraft;
+import net.minecraft.core.BlockPos;
+import net.minecraft.util.Tuple;
+import net.minecraft.world.entity.EquipmentSlot;
+import net.minecraft.world.level.block.Mirror;
+import net.minecraft.world.phys.AABB;
+
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.function.Supplier;
+
+import static com.ldtteam.structurize.items.ModItems.buildTool;
+
+/**
+ * Renders blueprints in the colony area.
+ */
+public class ColonyBlueprintRenderer
+{
+    /**
+     * An arbitrary not-null-but-invalid position.
+     */
+    private static final BlockPos INVALID_POS = BlockPos.ZERO.below(500);
+
+    /**
+     * The cached map of blueprints that are rendered.
+     */
+    private static Map<BlockPos, RenderData> blueprintCache = new HashMap<>();
+
+    private static BlockPos lastCacheRebuild = null;
+
+    /**
+     * Set of already requested structures.
+     */
+    public static Set<String> alreadyRequestedStructures = new HashSet<>();
+
+    /**
+     * Blueprints we're still loading.
+     */
+    private static Queue<PendingRenderData> pendingBlueprints = new LinkedList<>();
+
+    /**
+     * Rendering rules.  Order matters.
+     */
+    private static final List<IRenderBlueprintRule> renderRules = new ArrayList<>();
+
+    static
+    {
+        renderRules.add(new NearBuildPreview());
+        renderRules.add(new BuildGoggles());
+    }
+
+    /**
+     * Renders blueprints into the client.
+     *
+     * @param ctx rendering context
+     */
+    static void render(final WorldEventContext ctx)
+    {
+        if (!ctx.hasNearestColony())
+        {
+            blueprintCache.clear();
+            alreadyRequestedStructures.clear();
+            lastCacheRebuild = null;
+            return;
+        }
+
+        final List<IRenderBlueprintRule> activeRules = new ArrayList<>();
+        for (final IRenderBlueprintRule rule : renderRules)
+        {
+            if (rule.isEnabled(ctx))
+            {
+                activeRules.add(rule);
+            }
+        }
+
+        if (activeRules.isEmpty())
+        {
+            blueprintCache.clear();
+            alreadyRequestedStructures.clear();
+            lastCacheRebuild = null;
+            return;
+        }
+
+        if (lastCacheRebuild == null || Minecraft.getInstance().level.getGameTime() % 20 == 0)

```

Comment:
lastCacheRebuild is not used at all btw

imho this should be something like: if moved horizonzally (vertical is irrelevant?) for at least cca 8 to 10 blocks or else if 30 secs passed
since you want to prefer the move rule, the time rule is really minor probability - work orders don't change that often (afaik)

Reply:
This part was carried over from the previous implementation (which is hard to see on GH, although I did tag it as a rename on my end).

The pure time basis is sort of necessary, because this is also how often it processes its queue of pending blueprints -- it only updates at most one per timeslot, since that potentially includes a rotate/mirror operation (might be nice if that happened on the I/O thread instead....)  That part is also the same as before, the only thing I changed is that it can now process more than one queue slot if it didn't actually load a blueprint.

Reply:
loading queue should be disconnect from update rules, usually updating rules (= generating list of visible/loadable blueprints) requires (a lot?) searching and thus you dont want to run them so often - also they are position based, so it makes really small sense (for me) to make them time dependent (work order state change is only thing i can see happening - when you stand at the same pos)

Reply:
> which is hard to see on GH, although I did tag it as a rename on my end

afaik you cant tag something as rename, it should be just diff percent threshold

> That part is also the same as before, the only thing I changed is that it can now process more than one queue slot if it didn't actually load a blueprint.

well, Ray obviously (just like always :D ) doing things that are weird
the old impl was `if (lastCacheRebuild == null || !lastCacheRebuild.closerThan(activePosition, CACHE_RESET_RANGE))`
but Ray changed it to current one during blueprint loading update

Reply:
> afaik you cant tag something as rename, it should be just diff percent threshold

Yeah, that's what I meant, it was showing as a (delete+add =>) rename in the commit preview.  Looking at the history it seems it decided it wasn't after all, which is mildly annoying.  If I'd noticed that sooner I might have forced it via a helper commit, though it's not super important I guess.

Pushing a change in a minute that splits the cache regen from the queue processing, and cleans up a few other stray bits.  Slight regression in this version is that on login while wearing the goggles you won't see anything until you move past the threshold distance, though that's not too terrible.